### PR TITLE
Added exception translator specific mutex used with try_translate_exceptions

### LINKED ISSUE #5346: [BUG]: retranslating local exception with custom data class hangs in free-threading cpython
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### What version (or hash if on master) of pybind11 are you using?

2.13.5 and from source

### Problem description

I have a program stuck when retranslating local exception with custom data class using CPython with free-threading mode enabled.

This is due to `internals.mutex` being in the locked state when the callback is running. When using custom data, `all_type_info_get_cache` method is called internally which also tries to lock `internals.mutex`.

See the repro code below

### Reproducible example code

C++ test1.cpp

```c++
// https://pybind11.readthedocs.io/en/stable/basics.html
// clang++ -O3 -Wall -shared -std=c++11 -fPIC $(python -m pybind11 --includes) test1.cpp -o test1.so
// python -c "import test1; test1.check()"

#include <string>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>


namespace py = pybind11;

struct CustomData {
  CustomData(const std::string & a): a(a) {}
  std::string a;
};


struct CustomError {
  CustomError(const CustomData & message) : message(message) {}
  CustomData message;
};


PYBIND11_MODULE(test1, m, py::mod_gil_not_used()) {
  m.doc() = "test custom exception with free-threading";

  m.def("check", [](){
    auto d1 = CustomData("abc");
    throw CustomError(d1);
  });

  py::class_<CustomData>(m, "CustomData", py::module_local())
        .def(py::init<const std::string &>())
        .def_readwrite("a", &CustomData::a);
  py::register_local_exception_translator([](std::exception_ptr p) {
    try {
      if (p)
        std::rethrow_exception(p);
    } catch (const CustomError &e) {

      printf("Handle CustomError exception\n");
      auto mod = py::module_::import("exceptions1");
      py::object obj = mod.attr("CustomError");
      printf("Before the exception creation\n");

      // Here we can check that internals.mutex is locked: internals.mutex.mutex._bits == 1
      // obj(e.message) calls `all_type_info_get_cache` which would try to lock again internals

      // // If we unlock internals.mutex then obj(e.message) will work otherwise
      // // execution hangs.
      // auto &internals = py::detail::get_internals();
      // internals.mutex.unlock();

      py::object obj2 = obj(e.message);

      // We should lock again if we unlocked it
      // internals.mutex.lock();

      printf("After the exception creation\n");
      PyErr_SetObject(PyExc_Exception, obj2.ptr());

    }
  });
}
```

Python code: `exceptions1.py`
```python
class CustomError(Exception):
    def __init__(self, message):
        self.message = message
        super().__init__(message)

    def __str__(self):
        s = "[python]: " + self.message.a
        return s
```

Run:
```bash
python -c "import test1; test1.check()"
```
Output:
```bash
Handle CustomError exception
Before the exception creation
```

### Is this a regression? Put the last known working version here if it is.

Not a regression

### PR DESCRIPTION
**Pull request by vfdev-5**
## Description

- Added exception translator specific mutex in the internals to lock it when with `try_translate_exceptions` and `register_exception_translator`, `register_local_exception_translator`.

Fixes #5346

cc @colesbury

### CODE REVIEW

**Review by colesbury — COMMENTED** [review]
This looks like the right approach to me.

* I'd consider passing `internals.registered_exception_translators` to the callback and maybe rename it to `with_exception_translators` (instead of passing the internals struct).
* I think this is an ABI change for the free-threaded build due to the new mutex in `internals`, so I think `PYBIND11_INTERNALS_VERSION` needs to be incremented. It may make sense to only change it for the free-threaded build.
* A regression test derived from your bug report might be a good idea

**Comment by colesbury** [comment]
This LGTM, but needs review from the pybind11 maintainers

**Comment by vfdev-5** [comment]
Thanks for the feedback @colesbury !

@henryiii @rwgk could you please review this PR, thanks!

**Comment by rwgk** [comment]
(I'll look asap. I just changed companies and I'm still in the middle of getting set up. At the moment I don't even see when I'm getting tagged on GitHub, not sure why.)

File: include/pybind11/detail/internals.h
Comment (rwgk) [inline include/pybind11/detail/internals.h:43]: Unless #5296 is merged first, this will lead to incompatibility between extension modules compiled with or without `Py_GIL_DISABLED` defined.

I just only now realize ... Huston we have a problem?

We already added `pymutex mutex` to `struct internals` but didn't change `PYBIND11_INTERNALS_VERSION`. Ouch? Sorry I missed that before.

What do we want here? I'm very uncertain, @colesbury could you please advise?

```C++
struct internals {
#if PY_VERSION_HEX >= 0x030D0000
    pymutex mutex;
    pymutex exception_translator_mutex;
#endif
...
```

Could that work? The idea here is that `struct internals` is the same with and without `Py_GIL_DISABLED` defined.

Comment (henryiii) [reply include/pybind11/detail/internals.h:43]: Free-threading is experimental so changing the ABI before was fine. Now I think packages might be beginning to ship, so probably not such a great idea to change without a bump. But not fond of these confusing bumps. The ABI is currently a mess, by the way - the Windows only bump broke conda-forge and the only solution was to pin <2.12, which now is breaking packages (see the feedstock for more info).

Will the internals version matter after #5296? Could we just basically do away with the concept, and only do compiler/stdlib, etc?

Comment (rwgk) [reply include/pybind11/detail/internals.h:43]: > Could we just basically do away with the concept

Not entirely, **I think**. The internals versions still need to match for the inheritance functionality to work correctly. (I'd probably need a day or two to figure out conclusively if that's fixable or just fundamentally impossible.)

But the more I think about it, the more I'm getting worried about having a different `struct internals` with and without `Py_GIL_DISABLED` defined:

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/internals.h#L177-L180

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/internals.h#L185-L190

What happens if there is one extension built with the GIL enabled, another built with free-threading, and both are imported into the same process?

Currently we only have this:

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/common.h#L390-L405

Do we need to also check for `Py_GIL_DISABLED` compatibility there?

Or do we have to do something here?

* https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/internals.h#L324-L332

@colesbury

Comment (colesbury) [reply include/pybind11/detail/internals.h:43]: > What happens if there is one extension built with the GIL enabled, another built with free-threading, and both are imported into the same process?

You can't have extensions compiled with and without `Py_GIL_DISABLED` in the same process. An extension might require the GIL to be dynamically enabled, but that's different: that's a matter of calling (or not calling) `PyUnstable_Module_SetGIL()`. The extension will still need to be compiled with `Py_GIL_DISABLED` to be loaded in 3.13t.

In other words:

* `Py_GIL_DISABLED` is part of the extension ABI, similar to `Py_VERSION_HEX`
* It's part of the wheel tag (i.e., the "t" in `cp313-cp313t`), just like the version number

Comment (colesbury) [reply include/pybind11/detail/internals.h:43]: Or as @henryiii wrote in https://github.com/pybind/pybind11/pull/5148#discussion_r1619348108: "A free-threaded build can't talk to a non-free-threaded build".

Comment (colesbury) [reply include/pybind11/detail/internals.h:43]: @vfdev-5 that's not related to the `struct internals`, `PYBIND11_INTERNALS_VERSION`, or even free-threading.

Comment (vfdev-5) [reply include/pybind11/detail/internals.h:43]: I agree, I should have built using same python version, e.g. 3.13t and 3.13.
Doing so, above example is giving either a segfault:
```bash
$ gdb --args python3.13t main.py

Starting program: /usr/bin/python3.13t main.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7be1f0b in pybind11::dict::dict(pybind11::object&&) () from /pybind11_playground/ft_or_not_ft/test_gil.so
```
or 
```bash
$ python3.13 main.py
Traceback (most recent call last):
  File "/pybind11_playground/ft_or_not_ft/main.py", line 4, in <module>
    import test_ft
ImportError: /pybind11_playground/ft_or_not_ft/test_ft.so: undefined symbol: _Py_DecRefShared
```

If this is again unrelated sorry for the noise.

Comment (rwgk) [reply include/pybind11/detail/internals.h:43]: Thanks a lot! That really helps me understand the situation better. So I think what we need to do is work on this code:

https://github.com/pybind/pybind11/blob/5efc7439d43f3dbe8de2526df3436b8d99b10dab/include/pybind11/detail/common.h#L390-L405

We want to catch if there is a mismatch between the Python interpreter kind (gil, free-threading) and the way the extension was built.

**In a separate PR.**

Comment (rwgk) [reply include/pybind11/detail/internals.h:43]: > Now I think packages might be beginning to ship, so probably not such a great idea to change without a bump. But not fond of these confusing bumps.

Yeah ... "version bump chaos"

Trying to summarize how I understand the situation:

* The internals version wasn't changed for Python 3.13t — but that's ok because there were no released packages anyway that could have been built with a non-matching `struct internals`.

* Assuming there are released packages for 3.13t now, yes, it would indeed be safest to bump the internals version for 3.13t. — But pegging that on `Py_GIL_DISABLED` will make it very difficult for the general user community to reason about the internals version.

If this PR gets merged as is, we'll have internals version

* `4` for Python < 3.12 if `_MSC_VER` is not defined.
* `5` for Python 3.12, or 3.13, or if `_MSC_VER` is defined.
* `6` for Python 3.13t only.

Who can (or wants to) understand that? :-)

And then we also still have #5339 (smart_holder) and #5280 (native enum) that I'd really like to see merged.

I believe this is a situation where the maintainers need to step up to balance "making progress" with "minimizing confusion".

@henryiii wrote:

> Free-threading is experimental

That makes me think, let's not mess with the internals version in this PR:

* Maybe a small number of users **experimenting** will stumble,
* but a much larger number of **production** users will not have to understand the weirdness summarized above.

WDYT?

Looking beyond that question, I feel it'd be best to approach this strategically:

* Merge #5296 asap (only needs a small amount of work).
* Merge #5339, #5280 as closely together in time as possible and start the pybind11 v3 series, with internals version **`6`** unconditionally.

That's a discussion far beyond this PR, but I'm mentioning it here because it would be another reason to not mess with the internals version here.

Comment (colesbury) [reply include/pybind11/detail/internals.h:43]: > That makes me think, let's not mess with the internals version in this PR

I'm concerned that landing this PR without changing the version number will cause problems not just for users but also package maintainers (e.g., scipy, PyTorch) that use pybind11 and support free-threading. The failure modes for different internals can be complicated -- in this case, I think you run the risk of treating uninitialized memory like a mutex.

I think the approach in this PR is a good long term fix, but if the version bump is the issue we can first pursue a different approach that avoids changing the internals struct (and consider landing this change later when #5280 is ready).

Comment (rwgk) [reply include/pybind11/detail/internals.h:43]: First priority, I want to be helpful. Second minimize potential for confusion.

Unfortunately I'm technically somewhat handicapped at the moment (no Linux workstation), but let me try to get #5296 merged asap, with @henryiii's formal approval, to be sure we're on the same.

After that version bumps will have a lot less impact, and it won't matter as much that people fully understand the numbering.

File: tests/test_exceptions.cpp
Comment (rwgk) [inline tests/test_exceptions.cpp:408]: Could you please make this name more specific (e.g. `PythonMyException7_storage`)?

File: tests/test_exceptions.cpp
Comment (rwgk) [inline tests/test_exceptions.cpp:422]: For improved readability:
```
            auto exc_type = PythonMyException7_storage.get_stored();
            py::object exc_inst = exc_type(e.message);
```
Or similar.

**Review by rwgk — APPROVED** [review]
Looks good to me, based on @colesbury's review; I don't fully understand the mutex issue (#5346) TBH; I won't have enough free cycles for a while to really get my head around this.

PR #5296 was merged, which means we don't have to worry about the internals version bump so much.

**Comment by vfdev-5** [comment]
@rwgk I applied the suggestions to this PR. Anything else I need to do here or it is good to be merged?

**Review by rwgk — APPROVED** [review]

**Comment by rwgk** [comment]
Could you please add a Changlog entry to the PR description?

Please see other open PRs for the template.

**Comment by vfdev-5** [comment]
@rwgk I updated the description of the PR with a changelog entry, feel free to update it if needed or let me know how I can update it. Thanks!

**Pull request merged**
